integration/traffic/historicalspeed/ways_mapping.go

```diff
@@ -0,0 +1,228 @@
+	backwardItem, backwardOK := (*w)[-wayID]
+	if !forwardOK && !backwardOK {
+		return fmt.Errorf("no historical speed on the way")
+	}
+
+	tz, err := timezone.ParseTimezone(tzStr)
+	if err != nil {
+		return err
+	}
+	dst, err := timezone.ParseDaylightSaving(dstStr)
+	if err != nil {
+		return err
+	}
+
+	if forwardOK {
+		forwardItem.timezone = tz
+		forwardItem.daylightSaving = dst
+	}
+	if backwardOK {
+		backwardItem.timezone = tz
+		backwardItem.daylightSaving = dst
+	}
+
+	return nil
+}
+
+func (w *WaysMapping) loadFromSingleFile(filePath string) error {
+
+	f, err := os.Open(filePath)
+	defer f.Close()
+	if err != nil {
+		return err
+	}
+	glog.V(1).Infof("open %s succeed.\n", filePath)
+
+	r := csv.NewReader(f)
+	r.ReuseRecord = true
+
+	beforeLoadMappingCount := w.Count()
+	var count int // succeed parsed count
+	for {
+		record, err := r.Read()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			glog.Error(err)
+			return err
+		}
+
+		wayID, mapping, err := parseWaysMappingRecord(record)
+		if err != nil {
+			if count == 0 {
+				glog.V(2).Infof("Ignore head record %v due to parse failure: %v", record, err)
+			} else {
+				glog.Warningf("Parse record %v failed, err: %v", record, err)
+			}
+			continue
+		}
+
+		(*w)[wayID] = mapping
+		count++
+	}
+
+	glog.V(1).Infof("Loaded way2patterns mapping from file %s, count %d, total succeed parsed count %d", filePath, w.Count()-beforeLoadMappingCount, count)
+	return nil
+}
+
+func parseWaysMappingRecord(record []string) (int64, *mappingItem, error) {
+	if len(record) != fieldsPerCSVLine && len(record) != fieldsWithTimezoneDaylightSavingPerCSVLine {
+		return 0, nil, fmt.Errorf("expect %d or %d fields in csv record but got %d", fieldsPerCSVLine, fieldsWithTimezoneDaylightSavingPerCSVLine, len(record))
+	}
+
+	undirectedWayID, err := strconv.ParseUint(record[0], 10, 64)
+	if err != nil {
+		return 0, nil, fmt.Errorf("parse wayID from %s error: %v", record[0], err)
+	}
+	wayID := int64(undirectedWayID)
+
+	forward, err := strconv.ParseBool(record[1])
+	if err != nil {
+		return 0, nil, fmt.Errorf("parse wayID direction from %s error: %v", record[1], err)
+	}
+	if !forward {
+		wayID = -wayID
+	}
+
+	patternIDsRecord := record[2 : 2+daysPerWeek]
+	mapping := mappingItem{}
+
+	for i, v := range patternIDsRecord {
+		patternID, err := strconv.ParseUint(v, 10, 32)
+		if err != nil {
+			return 0, nil, fmt.Errorf("parse patternID from %s failed, err %v", v, err)
+		}
+		mapping.patternIDs[i] = uint32(patternID)
+	}
+
+	if len(record) == fieldsWithTimezoneDaylightSavingPerCSVLine { // prase timezoen and daylight_saving if exist
+
+		tz, err := timezone.ParseTimezone(record[9])
+		if err != nil {
+			return 0, nil, err
+		}
+		mapping.timezone = tz
+
+		dst, err := timezone.ParseDaylightSaving(record[10])
+		if err != nil {
+			return 0, nil, err
+		}
+		mapping.daylightSaving = dst
+	}
+
+	return wayID, &mapping, nil
+}
+
+func toWaysMappingRecord(wayID int64, item *mappingItem) []string {
```

Comment:
better to name as `convertToWaysMappingRecord`

When I saw `toWaysMappingRecord`, I tried to search `fromWaysMappingRecord`

Reply:
Changed it to `formatToWaysMappingRecord` to follow same `parse/format` convention from https://golang.org/pkg/strconv/